### DISCUSSION TITLE
Using `<M extends Model>` for the repository functions?

### DISCUSSION THREAD

**Discussion by MartinX3** [CodeDredd/pinia-orm#1829] [Ideas]
I think it would be great to use as example
```typescript
save(records: M[]): M[];
save(record: M): M;
```
instead of
```typescript
save(records: Element[]): M[];
save(record: Element): M;
```

So my IDE could autocomplete all the object attributes I'm writing down to as example save one or multiple objects in the database.

At the moment I do `useRepo(Bla, store).save([{...}, {...}, {...}] as Bla[])` to have auto completion and errors if I miss a non optional field.

**Comment by CodeDredd** [comment]
I think that's not a good idea, because maybe you want to pass an property to the save function which is not belonging to the model, but you want to react on it with a hook. If i type check the save function then this is no longer possible.
Another point are relations. It would be really hard to type check an big dataset with relations.

**Reply by MartinX3** [reply]
Oh, sad.
Do you have an suggestion to make sure I don't miss mandatory fields and maybe get an autocompletion?

And what would be a use case to save fields which I didn't specify in the entity model?
I think I never used it in backend programming with spring boot and jpa database connections.

**Reply by CodeDredd** [reply]
> Do you have an suggestion to make sure I don't miss mandatory fields and maybe get an autocompletion?

Use the hooks for that. E.g. the `creating` hook and return false if one mandtory prop is missing with an error message. That way it's not saved and you know what to look at.
> And what would be a use case to save fields which I didn't specify in the entity model?

It's not about saving. The props could be used e.g. in the `creating` hook where you check maybe a specific feature flag which triggers something but is not important to be saved.

**Reply by MartinX3** [reply]
Ahhhhh, thank you!

**Reply by CodeDredd** [reply]
This brings me actually to the idea to introduce a `required` flag for fields which if not set are throwing an error and canceling the save.

**Reply by MartinX3** [reply]
I thought non optional/null fields are mandatory by default.

**Reply by CodeDredd** [reply]
true....but they are only throwing a warning. I think a feature flag is needed if you want to throw this as error and canceling the action.
